pkg/resourcekeeper/statekeep.go

```diff
@@ -67,3 +80,160 @@ func (h *resourceKeeper) StateKeep(ctx context.Context) error {
 	}
 	return nil
 }
+
+// todo applyonce处理*的情况, 测试只有 component和 resource 的情况
+// applyOnceManifest will generate manifest with applyOnceStrategy
+func applyOnceManifest(ctx context.Context, h *resourceKeeper, manifest *unstructured.Unstructured, applyOnceStrategy *v1alpha1.ApplyOnceStrategy) (*unstructured.Unstructured, error) {
+	un := new(unstructured.Unstructured)
+	un.SetAPIVersion(manifest.GetAPIVersion())
+	un.SetKind(manifest.GetKind())
+	err := h.Get(ctx, types.NamespacedName{Name: manifest.GetName(), Namespace: manifest.GetNamespace()}, un)
+	if err != nil {
+		return nil, err
+	}
+
+	for _, path := range applyOnceStrategy.Path {
+		if path == "*" {
+			manifest = un.DeepCopy()
+			break
+		}
+		value, err := getNestedValueWithSlice(un, path)
+		if err != nil {
+			return nil, err
+		}
+		manifest, err = setNestedValueWithSlice(manifest, path, value)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	return manifest, nil
+}
+
+// getNestedValueWithSlice get the value of unstructured.Unstructured via the path, like spec.template.spec.containers[0].image
+func getNestedValueWithSlice(manifest *unstructured.Unstructured, path string) (interface{}, error) {
+label:
+	prefix, index, suffix := splitPathWithIndex(path)
+	// 是 slice
+	if index != -1 {
+		dotPath := strings.Split(prefix, ".")
+		nestedSlice, found, err := unstructured.NestedSlice(manifest.UnstructuredContent(), dotPath...)
+		if err != nil {
+			return nil, err
+		}
+		if !found || len(nestedSlice) <= index {
+			return nil, errors.Errorf("the path(%s) was not found ", path)
+		}
+		manifest.Object = nestedSlice[index].(map[string]interface{})
+		path = suffix
+		goto label
+	}
+	// 不是 slice
+	if index == -1 {
+		dotPath := strings.Split(prefix, ".")
+		nestedField, found, err := unstructured.NestedFieldCopy(manifest.UnstructuredContent(), dotPath...)
+		if err != nil {
+			return nil, err
+		}
+		if !found {
+			return nil, errors.Errorf("the path(%s) was not found ", path)
+		}
+		return nestedField, nil
+	}
+	return nil, nil
+}
+
+// setNestedValueWithSlice set the value to unstructured.Unstructured via the path, like spec.template.spec.containers[0].image
+func setNestedValueWithSlice(manifest *unstructured.Unstructured, path string, value interface{}) (*unstructured.Unstructured, error) {
```

Comment:
`github.com/crossplane/crossplane-runtime/pkg/fieldpath`

Please check if this library can help, we have somthing code related with it, you can search with this package.

It works like this:

```
err = fieldpath.Pave(manifest).SetValue(path, value)
```